File: benchmarks/tsb/bench_dataframe_rolling_apply_fn.ts
/**
 * Benchmark: standalone dataFrameRollingApply — apply a custom function over each column's rolling window.
 * Outputs JSON: {"function": "dataframe_rolling_apply_fn", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { DataFrame, dataFrameRollingApply } from "../../src/index.ts";

const ROWS = 5_000;
const WINDOW = 10;
const WARMUP = 3;
const ITERATIONS = 10;

const df = DataFrame.fromColumns({
  a: Array.from({ length: ROWS }, (_, i) => Math.sin(i * 0.01)),
  b: Array.from({ length: ROWS }, (_, i) => Math.cos(i * 0.02)),
  c: Array.from({ length: ROWS }, (_, i) => (i % 100) * 0.5),
});

const rangeFn = (vals: readonly number[]) => {
  let mn = vals[0] ?? 0;
  let mx = vals[0] ?? 0;
  for (const v of vals) {
    if (v < mn) mn = v;
    if (v > mx) mx = v;
  }
  return mx - mn;
};

for (let i = 0; i < WARMUP; i++) {
  dataFrameRollingApply(df, WINDOW, rangeFn);
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  dataFrameRollingApply(df, WINDOW, rangeFn);
}
const total = performance.now() - start;

console.log(
  JSON.stringify({
    function: "dataframe_rolling_apply_fn",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);

File: benchmarks/pandas/bench_dataframe_rolling_apply_fn.py
"""Benchmark: DataFrame rolling apply with a custom range function per column."""
import json, time
import numpy as np
import pandas as pd

ROWS = 5_000
WINDOW = 10
WARMUP = 3
ITERATIONS = 10

a = np.sin(np.arange(ROWS) * 0.01)
b = np.cos(np.arange(ROWS) * 0.02)
c = (np.arange(ROWS) % 100) * 0.5
df = pd.DataFrame({"a": a, "b": b, "c": c})

range_fn = lambda w: np.max(w) - np.min(w)

for _ in range(WARMUP):
    df.rolling(WINDOW).apply(range_fn, raw=True)

start = time.perf_counter()
for _ in range(ITERATIONS):
    df.rolling(WINDOW).apply(range_fn, raw=True)
total = (time.perf_counter() - start) * 1000

print(json.dumps({
    "function": "dataframe_rolling_apply_fn",
    "mean_ms": total / ITERATIONS,
    "iterations": ITERATIONS,
    "total_ms": total,
}))
